S3 Compatibility failures.

Made a small s3Verify tool by turning the existing `server_test` and `server_xl_test` with a small change. 
Here are the changes to be made to replicate the test. 

```
@@ -43,6 +44,9 @@ var _ = Suite(&MyAPISuite{})
 // Starting the Test server with temporary FS backend.
 func (s *MyAPISuite) SetUpSuite(c *C) {
        s.testServer = StartTestServer(c, "FS")
+      s.testServer.Server.URL = "https://s3.amazonaws.com"
+      s.testServer.AccessKey = "<S3-ACCESS-KEY" 
+       s.testServer.SecretKey = "<s3-SECRET-KEY>"
 }

 // Called implicitly by "gopkg.in/check.v1" after all tests are run.
diff --git a/server_utils_test.go b/server_utils_test.go
index e95b6ee..22d1dc4 100644
--- a/server_utils_test.go
+++ b/server_utils_test.go
@@ -82,11 +82,11 @@ func getFreePort() int {

 func verifyError(c *C, response *http.Response, code, description string, statusCode int) {
        data, err := ioutil.ReadAll(response.Body)
-       c.Assert(err, IsNil)
+       c.Check(err, IsNil)
        errorResponse := APIErrorResponse{}
        err = xml.Unmarshal(data, &errorResponse)
-       c.Assert(err, IsNil)
-       c.Assert(errorResponse.Code, Equals, code)
-       c.Assert(errorResponse.Message, Equals, description)
-       c.Assert(response.StatusCode, Equals, statusCode)
+       c.Check(err, IsNil)
+       c.Check(errorResponse.Code, Equals, code)
+       c.Check(errorResponse.Message, Equals, description)
+       c.Check(response.StatusCode, Equals, statusCode)
 }
diff --git a/server_xl_test.go b/server_xl_test.go
index 4814cdc..c560e21 100644
--- a/server_xl_test.go
+++ b/server_xl_test.go
@@ -45,6 +45,9 @@ var _ = Suite(&MyAPIXLSuite{})
 // Starting the Test server with temporary XL backend.
 func (s *MyAPIXLSuite) SetUpSuite(c *C) {
        s.testServer = StartTestServer(c, "XL")
+       s.testServer.Server.URL = "https://s3.amazonaws.com"
+       s.testServer.AccessKey = "<S3-ACCESS-KEY>" 
+        s.testServer.SecretKey = "<S3-SECRET>"
 }

```
## Expected behaviour

Since the tests pass for Minio sever, its expected to pass for Amazon s3 too if we are 100% s3 compatible. 
## Actual behaviour

Around 32 tests failed. Here is the log of failed tests. 
## Steps to reproduce the behaviour
## Minio version
- (paste output of `minio version`)

```

----------------------------------------------------------------------
FAIL: server_test.go:292: MyAPISuite.TestBucket

server_test.go:300:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 409
... expected int = 200


----------------------------------------------------------------------
FAIL: server_test.go:68: MyAPISuite.TestBucketPolicy

server_test.go:113:
    c.Assert(response.StatusCode, Equals, http.StatusNoContent)
... obtained int = 403
... expected int = 204


----------------------------------------------------------------------
FAIL: server_test.go:727: MyAPISuite.TestContentTypePersists

server_test.go:753:
    c.Assert(response.Header.Get("Content-Type"), Equals, "image/png")
... obtained string = "binary/octet-stream"
... expected string = "image/png"


----------------------------------------------------------------------
FAIL: server_test.go:498: MyAPISuite.TestCopyObject

server_test.go:524:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 403
... expected int = 200


----------------------------------------------------------------------
FAIL: server_test.go:151: MyAPISuite.TestDeleteBucket

server_test.go:159:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 409
... expected int = 200


----------------------------------------------------------------------
FAIL: server_test.go:902: MyAPISuite.TestGetObjectErrors

server_test.go:910:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket does not exist."

server_test.go:936:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = "NoSuchBucket"
... expected string = "InvalidBucketName"

server_test.go:936:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket is not valid."

server_test.go:936:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 404
... expected int = 400


----------------------------------------------------------------------
FAIL: server_test.go:940: MyAPISuite.TestGetObjectRangeErrors

server_test.go:968:
    verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
server_utils_test.go:88:
    c.Check(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"EOF"} ("EOF")

server_test.go:968:
    verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = ""
... expected string = "InvalidRange"

server_test.go:968:
    verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = ""
... expected string = "The requested range cannot be satisfied."

server_test.go:968:
    verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 200
... expected int = 416


----------------------------------------------------------------------
FAIL: server_test.go:618: MyAPISuite.TestHeadOnObject

server_test.go:655:
    c.Assert(response.StatusCode, Equals, http.StatusNotModified)
... obtained int = 200
... expected int = 304


----------------------------------------------------------------------
FAIL: server_test.go:452: MyAPISuite.TestHeader

server_test.go:461:
    verifyError(c, response, "NoSuchKey", "The specified key does not exist.", http.StatusNotFound)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = "AccessDenied"
... expected string = "NoSuchKey"

server_test.go:461:
    verifyError(c, response, "NoSuchKey", "The specified key does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "Access Denied"
... expected string = "The specified key does not exist."

server_test.go:461:
    verifyError(c, response, "NoSuchKey", "The specified key does not exist.", http.StatusNotFound)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 403
... expected int = 404


----------------------------------------------------------------------
FAIL: server_test.go:829: MyAPISuite.TestListObjectsHandlerErrors

server_test.go:837:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = "NoSuchBucket"
... expected string = "InvalidBucketName"

server_test.go:837:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket is not valid."

server_test.go:837:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 404
... expected int = 400

server_test.go:846:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket does not exist."

server_test.go:863:
    verifyError(c, response, "InvalidArgument", "Argument maxKeys must be an integer between 0 and 2147483647.", http.StatusBadRequest)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "Argument maxKeys must be an integer between 0 and 2147483647"
... expected string = "Argument maxKeys must be an integer between 0 and 2147483647."


----------------------------------------------------------------------
FAIL: server_test.go:606: MyAPISuite.TestNotBeAbleToCreateObjectInNonexistentBucket

server_test.go:615:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket does not exist."


----------------------------------------------------------------------
FAIL: server_test.go:441: MyAPISuite.TestNotImplemented

server_test.go:449:
    c.Assert(response.StatusCode, Equals, http.StatusNotImplemented)
... obtained int = 403
... expected int = 501


----------------------------------------------------------------------
FAIL: server_test.go:312: MyAPISuite.TestObjectGet

server_test.go:321:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 409
... expected int = 200


----------------------------------------------------------------------
FAIL: server_test.go:1143: MyAPISuite.TestObjectMultipartList

server_test.go:1200:
    verifyError(c, response4, "InvalidArgument", "Argument maxParts must be an integer between 1 and 10000.", http.StatusBadRequest)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "Argument max-parts must be an integer between 0 and 2147483647"
... expected string = "Argument maxParts must be an integer between 1 and 10000."


----------------------------------------------------------------------
FAIL: server_test.go:866: MyAPISuite.TestPutBucketErrors

server_test.go:883:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 409
... expected int = 200


----------------------------------------------------------------------
FAIL: server_test.go:576: MyAPISuite.TestPutObjectLongName

server_test.go:603:
    c.Assert(response.StatusCode, Equals, http.StatusNotFound)
... obtained int = 501
... expected int = 404


----------------------------------------------------------------------
FAIL: server_xl_test.go:304: MyAPIXLSuite.TestBucket

server_xl_test.go:312:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 409
... expected int = 200


----------------------------------------------------------------------
FAIL: server_xl_test.go:69: MyAPIXLSuite.TestBucketPolicy

server_xl_test.go:114:
    c.Assert(response.StatusCode, Equals, http.StatusNoContent)
... obtained int = 403
... expected int = 204


----------------------------------------------------------------------
FAIL: server_xl_test.go:512: MyAPIXLSuite.TestCopyObject

server_xl_test.go:539:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 403
... expected int = 200


----------------------------------------------------------------------
FAIL: server_xl_test.go:152: MyAPIXLSuite.TestDeleteBucket

server_xl_test.go:160:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 409
... expected int = 200


----------------------------------------------------------------------
FAIL: server_xl_test.go:1133: MyAPIXLSuite.TestGetObjectErrors

server_xl_test.go:1141:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:88:
    c.Check(err, IsNil)
... value xml.UnmarshalError = "expected element type <Error> but have <ListBucketResult>"

server_xl_test.go:1141:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = ""
... expected string = "NoSuchBucket"

server_xl_test.go:1141:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = ""
... expected string = "The specified bucket does not exist."

server_xl_test.go:1141:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 200
... expected int = 404

server_xl_test.go:1167:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = "NoSuchBucket"
... expected string = "InvalidBucketName"

server_xl_test.go:1167:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket is not valid."

server_xl_test.go:1167:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 404
... expected int = 400


----------------------------------------------------------------------
FAIL: server_xl_test.go:943: MyAPIXLSuite.TestGetObjectLarge10MiB

server_xl_test.go:952:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 409
... expected int = 200


----------------------------------------------------------------------
FAIL: server_xl_test.go:1171: MyAPIXLSuite.TestGetObjectRangeErrors

server_xl_test.go:1199:
    verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
server_utils_test.go:88:
    c.Check(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"EOF"} ("EOF")

server_xl_test.go:1199:
    verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = ""
... expected string = "InvalidRange"

server_xl_test.go:1199:
    verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = ""
... expected string = "The requested range cannot be satisfied."

server_xl_test.go:1199:
    verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 200
... expected int = 416


----------------------------------------------------------------------
FAIL: server_xl_test.go:648: MyAPIXLSuite.TestHeadOnObject

server_xl_test.go:685:
    c.Assert(response.StatusCode, Equals, http.StatusNotModified)
... obtained int = 200
... expected int = 304


----------------------------------------------------------------------
FAIL: server_xl_test.go:465: MyAPIXLSuite.TestHeader

server_xl_test.go:474:
    verifyError(c, response, "NoSuchKey", "The specified key does not exist.", http.StatusNotFound)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = "AccessDenied"
... expected string = "NoSuchKey"

server_xl_test.go:474:
    verifyError(c, response, "NoSuchKey", "The specified key does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "Access Denied"
... expected string = "The specified key does not exist."

server_xl_test.go:474:
    verifyError(c, response, "NoSuchKey", "The specified key does not exist.", http.StatusNotFound)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 403
... expected int = 404


----------------------------------------------------------------------
FAIL: server_xl_test.go:870: MyAPIXLSuite.TestListObjectsHandlerErrors

server_xl_test.go:878:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = "NoSuchBucket"
... expected string = "InvalidBucketName"

server_xl_test.go:878:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket is not valid."

server_xl_test.go:878:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 404
... expected int = 400

server_xl_test.go:887:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:88:
    c.Check(err, IsNil)
... value xml.UnmarshalError = "expected element type <Error> but have <ListBucketResult>"

server_xl_test.go:887:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:89:
    c.Check(errorResponse.Code, Equals, code)
... obtained string = ""
... expected string = "NoSuchBucket"

server_xl_test.go:887:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = ""
... expected string = "The specified bucket does not exist."

server_xl_test.go:887:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:91:
    c.Check(response.StatusCode, Equals, statusCode)
... obtained int = 200
... expected int = 404

server_xl_test.go:904:
    verifyError(c, response, "InvalidArgument", "Argument maxKeys must be an integer between 0 and 2147483647.", http.StatusBadRequest)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "Argument maxKeys must be an integer between 0 and 2147483647"
... expected string = "Argument maxKeys must be an integer between 0 and 2147483647."


----------------------------------------------------------------------
FAIL: server_xl_test.go:636: MyAPIXLSuite.TestNotBeAbleToCreateObjectInNonexistentBucket

server_xl_test.go:645:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket does not exist."


----------------------------------------------------------------------
FAIL: server_xl_test.go:454: MyAPIXLSuite.TestNotImplemented

server_xl_test.go:462:
    c.Assert(response.StatusCode, Equals, http.StatusNotImplemented)
... obtained int = 403
... expected int = 501


----------------------------------------------------------------------
FAIL: server_xl_test.go:324: MyAPIXLSuite.TestObject

server_xl_test.go:333:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 409
... expected int = 200


----------------------------------------------------------------------
FAIL: server_xl_test.go:1388: MyAPIXLSuite.TestObjectMultipartList

server_xl_test.go:1445:
    verifyError(c, response4, "InvalidArgument", "Argument maxParts must be an integer between 1 and 10000.", http.StatusBadRequest)
server_utils_test.go:90:
    c.Check(errorResponse.Message, Equals, description)
... obtained string = "Argument max-parts must be an integer between 0 and 2147483647"
... expected string = "Argument maxParts must be an integer between 1 and 10000."


----------------------------------------------------------------------
FAIL: server_xl_test.go:907: MyAPIXLSuite.TestPutBucketErrors

server_xl_test.go:924:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 409
... expected int = 200


----------------------------------------------------------------------
FAIL: server_xl_test.go:590: MyAPIXLSuite.TestPutObjectLongName

server_xl_test.go:617:
    c.Assert(response.StatusCode, Equals, http.StatusNotFound)
... obtained int = 501
... expected int = 404

--- FAIL: Test (77.35s)
FAIL
exit status 1
FAIL    github.com/minio/minio  80.181s

```
## System information

OSX ElCaptain.

Here are failures with the latest run

```

----------------------------------------------------------------------
FAIL: server_xl_test.go:75: MyAPIXLSuite.TestBucketPolicy

server_xl_test.go:150:
    // Verify if downloaded policy matches with previousy uploaded.
    c.Assert(bytes.Equal([]byte(bucketPolicyStr), bucketPolicyReadBuf), Equals, true)
... obtained bool = false
... expected bool = true


----------------------------------------------------------------------
FAIL: server_xl_test.go:661: MyAPIXLSuite.TestCopyObject

server_xl_test.go:700:
    c.Assert(response.StatusCode, Equals, http.StatusOK)
... obtained int = 403
... expected int = 200


----------------------------------------------------------------------
FAIL: server_xl_test.go:1464: MyAPIXLSuite.TestGetObjectErrors

server_xl_test.go:1474:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Assert(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket does not exist."


----------------------------------------------------------------------
FAIL: server_xl_test.go:1504: MyAPIXLSuite.TestGetObjectRangeErrors

server_xl_test.go:1536:
    verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
server_utils_test.go:88:
    c.Assert(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"EOF"} ("EOF")


----------------------------------------------------------------------
FAIL: server_xl_test.go:836: MyAPIXLSuite.TestHeadOnObject

server_xl_test.go:877:
    c.Assert(response.StatusCode, Equals, http.StatusNotModified)
... obtained int = 200
... expected int = 304


----------------------------------------------------------------------
FAIL: server_xl_test.go:601: MyAPIXLSuite.TestHeader

server_xl_test.go:613:
    // asserting for the expected error response.
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Assert(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket does not exist."


----------------------------------------------------------------------
FAIL: server_xl_test.go:1084: MyAPIXLSuite.TestListObjectsHandlerErrors

server_xl_test.go:1094:
    verifyError(c, response, "InvalidBucketName", "The specified bucket is not valid.", http.StatusBadRequest)
server_utils_test.go:89:
    c.Assert(errorResponse.Code, Equals, code)
... obtained string = "NoSuchBucket"
... expected string = "InvalidBucketName"


----------------------------------------------------------------------
FAIL: server_xl_test.go:821: MyAPIXLSuite.TestNotBeAbleToCreateObjectInNonexistentBucket

server_xl_test.go:832:
    verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
server_utils_test.go:90:
    c.Assert(errorResponse.Message, Equals, description)
... obtained string = "The specified bucket does not exist"
... expected string = "The specified bucket does not exist."


----------------------------------------------------------------------
FAIL: server_xl_test.go:587: MyAPIXLSuite.TestNotImplemented

server_xl_test.go:597:
    c.Assert(response.StatusCode, Equals, http.StatusNotImplemented)
... obtained int = 404
... expected int = 501


----------------------------------------------------------------------
FAIL: server_xl_test.go:1740: MyAPIXLSuite.TestObjectMultipartList

server_xl_test.go:1801:
    verifyError(c, response4, "InvalidArgument", "Argument maxParts must be an integer between 1 and 10000.", http.StatusBadRequest)
server_utils_test.go:90:
    c.Assert(errorResponse.Message, Equals, description)
... obtained string = "Argument max-parts must be an integer between 0 and 2147483647"
... expected string = "Argument maxParts must be an integer between 1 and 10000."


----------------------------------------------------------------------
FAIL: server_xl_test.go:1125: MyAPIXLSuite.TestPutBucketErrors

server_xl_test.go:1155:
    verifyError(c, response, "BucketAlreadyOwnedByYou", "Your previous request to create the named bucket succeeded and you already own it.",
        http.StatusConflict)
server_utils_test.go:88:
    c.Assert(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"EOF"} ("EOF")


----------------------------------------------------------------------
FAIL: server_xl_test.go:765: MyAPIXLSuite.TestPutObjectLongName

server_xl_test.go:797:
    c.Assert(response.StatusCode, Equals, http.StatusNotFound)
... obtained int = 501
... expected int = 404

--- FAIL: Test (55.44s)
FAIL
exit status 1
FAIL    github.com/minio/minio  56.633s

```

--- a/api-errors.go
+++ b/api-errors.go
@@ -133,17 +133,17 @@ var errorCodeResponse = map[APIErrorCode]APIError{
 	},
 	ErrInvalidMaxUploads: {
 		Code:           "InvalidArgument",
-		Description:    "Argument maxUploads must be an integer between 0 and 2147483647.",
+		Description:    "Argument max-uploads must be an integer between 0 and 2147483647",
 		HTTPStatusCode: http.StatusBadRequest,
 	},
 	ErrInvalidMaxKeys: {
 		Code:           "InvalidArgument",
-		Description:    "Argument maxKeys must be an integer between 0 and 2147483647.",
+		Description:    "Argument maxKeys must be an integer between 0 and 2147483647",
 		HTTPStatusCode: http.StatusBadRequest,
 	},
 	ErrInvalidMaxParts: {
 		Code:           "InvalidArgument",
-		Description:    "Argument maxParts must be an integer between 1 and 10000.",
+		Description:    "Argument max-parts must be an integer between 0 and 2147483647",
 		HTTPStatusCode: http.StatusBadRequest,
 	},
 	ErrInvalidPartNumberMarker: {
@@ -208,7 +208,7 @@ var errorCodeResponse = map[APIErrorCode]APIError{
 	},
 	ErrInvalidRange: {
 		Code:           "InvalidRange",
-		Description:    "The requested range cannot be satisfied.",
+		Description:    "The requested range is not satisfiable",
 		HTTPStatusCode: http.StatusRequestedRangeNotSatisfiable,
 	},
 	ErrMalformedXML: {
@@ -233,7 +233,7 @@ var errorCodeResponse = map[APIErrorCode]APIError{
 	},
 	ErrNoSuchBucket: {
 		Code:           "NoSuchBucket",
-		Description:    "The specified bucket does not exist.",
+		Description:    "The specified bucket does not exist",
 		HTTPStatusCode: http.StatusNotFound,
 	},
 	ErrNoSuchBucketPolicy: {
@@ -253,7 +253,7 @@ var errorCodeResponse = map[APIErrorCode]APIError{
 	},
 	ErrNotImplemented: {
 		Code:           "NotImplemented",
-		Description:    "A header you provided implies functionality that is not implemented.",
+		Description:    "A header you provided implies functionality that is not implemented",
 		HTTPStatusCode: http.StatusNotImplemented,
 	},
 	ErrPreconditionFailed: {
@@ -461,7 +461,7 @@ func toAPIErrorCode(err error) (apiErr APIErrorCode) {
 	case ObjectNotFound:
 		apiErr = ErrNoSuchKey
 	case ObjectNameInvalid:
-		apiErr = ErrNoSuchKey
+		apiErr = ErrNotImplemented
 	case InvalidUploadID:
 		apiErr = ErrNoSuchUpload
 	case InvalidPart:

--- a/server_test.go
+++ b/server_test.go
@@ -390,8 +390,11 @@ func (s *TestSuiteCommon) TestEmptyObject(c *C) {
 }
 
 func (s *TestSuiteCommon) TestBucket(c *C) {
-	request, err := newTestRequest("PUT", s.testServer.Server.URL+"/bucket",
-		0, nil, s.testServer.AccessKey, s.testServer.SecretKey)
+	// generate a random bucket name.
+	bucketName := getRandomBucketName()
+
+	request, err := newTestRequest("PUT", getMakeBucketURL(s.endPoint, bucketName),
+		0, nil, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
 
 	client := http.Client{}
@@ -399,8 +402,8 @@ func (s *TestSuiteCommon) TestBucket(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(response.StatusCode, Equals, http.StatusOK)
 
-	request, err = newTestRequest("HEAD", s.testServer.Server.URL+"/bucket",
-		0, nil, s.testServer.AccessKey, s.testServer.SecretKey)
+	request, err = newTestRequest("HEAD", getMakeBucketURL(s.endPoint, bucketName),
+		0, nil, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
 
 	client = http.Client{}
@@ -580,9 +583,9 @@ func (s *TestSuiteCommon) TestMultipleObjects(c *C) {
 	c.Assert(true, Equals, bytes.Equal(responseBody, []byte("hello three")))
 }
 
-// TestNotImplemented - Validates response for obtaining policy on an non-existent bucket and object.
+// TestNotImplemented - validates if object policy is implemented, should return 'NotImplemented'.
 func (s *TestSuiteCommon) TestNotImplemented(c *C) {
-	// generate a random bucket name.
+	// Generate a random bucket name.
 	bucketName := getRandomBucketName()
 	request, err := newTestRequest("GET", s.endPoint+"/"+bucketName+"/object?policy",
 		0, nil, s.accessKey, s.secretKey)
@@ -607,7 +610,7 @@ func (s *TestSuiteCommon) TestHeader(c *C) {
 	response, err := client.Do(request)
 	c.Assert(err, IsNil)
 	// asserting for the expected error response.
-	verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
+	verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist", http.StatusNotFound)
 }
 
 func (s *TestSuiteCommon) TestPutBucket(c *C) {
@@ -685,8 +688,7 @@ func (s *TestSuiteCommon) TestCopyObject(c *C) {
 	// creating HTTP request for uploading the object.
 	request, err = newTestRequest("PUT", getPutObjectURL(s.endPoint, bucketName, objectName2),
 		0, nil, s.accessKey, s.secretKey)
-	// setting the "X-Amz-Copy-Source" to allow copying the content of
-	// previously uploaded object.
+	// setting the "X-Amz-Copy-Source" to allow copying the content of previously uploaded object.
 	request.Header.Set("X-Amz-Copy-Source", "/"+bucketName+"/"+objectName)
 	c.Assert(err, IsNil)
 	// execute the HTTP request.
@@ -814,7 +816,7 @@ func (s *TestSuiteCommon) TestPutObjectLongName(c *C) {
 
 	response, err = client.Do(request)
 	c.Assert(err, IsNil)
-	c.Assert(response.StatusCode, Equals, http.StatusNotFound)
+	verifyError(c, response, "NotImplemented", "A header you provided implies functionality that is not implemented", http.StatusNotImplemented)
 }
 
 // TestNotBeAbleToCreateObjectInNonexistentBucket - Validates the error response
@@ -836,7 +838,7 @@ func (s *TestSuiteCommon) TestNotBeAbleToCreateObjectInNonexistentBucket(c *C) {
 	response, err := client.Do(request)
 	c.Assert(err, IsNil)
 	// Assert the response error message.
-	verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist.", http.StatusNotFound)
+	verifyError(c, response, "NoSuchBucket", "The specified bucket does not exist", http.StatusNotFound)
 }
 
 // TestHeadOnObjectLastModified - Asserts response for HEAD on an object.
@@ -889,16 +891,16 @@ func (s *TestSuiteCommon) TestHeadOnObjectLastModified(c *C) {
 	c.Assert(err, IsNil)
 
 	// make HTTP request to obtain object info.
-	// But this time set the "If-Modified-Since" header to be a minute more than the actual
+	// But this time set the "If-Modified-Since" header to be 10 minute more than the actual
 	// last modified time of the object.
 	request, err = newTestRequest("HEAD", getHeadObjectURL(s.endPoint, bucketName, objectName),
 		0, nil, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
-	request.Header.Set("If-Modified-Since", t.Add(1*time.Minute).UTC().Format(http.TimeFormat))
+	request.Header.Set("If-Modified-Since", t.Add(10*time.Minute).UTC().Format(http.TimeFormat))
 	response, err = client.Do(request)
 	c.Assert(err, IsNil)
-	// Since the "If-Modified-Since" header was ahead in time compared to the actual modified time of the object
-	// expecting the response status to be http.StatusNotModified.
+	// Since the "If-Modified-Since" header was ahead in time compared to the actual
+	// modified time of the object expecting the response status to be http.StatusNotModified.
 	c.Assert(response.StatusCode, Equals, http.StatusNotModified)
 
 	// Again, obtain the object info.
@@ -907,7 +909,7 @@ func (s *TestSuiteCommon) TestHeadOnObjectLastModified(c *C) {
 	request, err = newTestRequest("HEAD", getHeadObjectURL(s.endPoint, bucketName, objectName),
 		0, nil, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
-	request.Header.Set("If-Unmodified-Since", t.Add(-1*time.Minute).UTC().Format(http.TimeFormat))
+	request.Header.Set("If-Unmodified-Since", t.Add(-10*time.Minute).UTC().Format(http.TimeFormat))
 	response, err = client.Do(request)
 	c.Assert(err, IsNil)
 	c.Assert(response.StatusCode, Equals, http.StatusPreconditionFailed)
@@ -955,7 +957,7 @@ func (s *TestSuiteCommon) TestContentTypePersists(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(response.StatusCode, Equals, http.StatusOK)
 
-	// Uploading a new object with Content-Type  "application/zip".
+	// Uploading a new object with Content-Type "image/png".
 	// content for the object to be uploaded.
 	buffer1 := bytes.NewReader([]byte("hello world"))
 	objectName := "test-object.png"
@@ -963,8 +965,7 @@ func (s *TestSuiteCommon) TestContentTypePersists(c *C) {
 	request, err = newTestRequest("PUT", getPutObjectURL(s.endPoint, bucketName, objectName),
 		int64(buffer1.Len()), buffer1, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
-
-	delete(request.Header, "Content-Type")
+	request.Header.Set("Content-Type", "image/png")
 
 	client = http.Client{}
 	// execute the HTTP request for object upload.
@@ -1001,12 +1002,9 @@ func (s *TestSuiteCommon) TestContentTypePersists(c *C) {
 	buffer2 := bytes.NewReader([]byte("hello world"))
 	request, err = newTestRequest("PUT", getPutObjectURL(s.endPoint, bucketName, objectName),
 		int64(buffer2.Len()), buffer2, s.accessKey, s.secretKey)
-	// deleting the old header value.
-	delete(request.Header, "Content-Type")
-	// setting the request header to be application/json.
-	request.Header.Add("Content-Type", "application/json")
-
 	c.Assert(err, IsNil)
+	// setting the request header to be application/json.
+	request.Header.Set("Content-Type", "application/json")
 
 	// Execute the HTTP request to upload the object.
 	response, err = client.Do(request)
@@ -1039,8 +1037,10 @@ func (s *TestSuiteCommon) TestContentTypePersists(c *C) {
 // By setting the Range header, A request to send specific bytes range of data from an
 // already uploaded object can be done.
 func (s *TestSuiteCommon) TestPartialContent(c *C) {
-	request, err := newTestRequest("PUT", s.testServer.Server.URL+"/partial-content",
-		0, nil, s.testServer.AccessKey, s.testServer.SecretKey)
+	bucketName := getRandomBucketName()
+
+	request, err := newTestRequest("PUT", getMakeBucketURL(s.endPoint, bucketName),
+		0, nil, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
 
 	client := http.Client{}
@@ -1049,8 +1049,8 @@ func (s *TestSuiteCommon) TestPartialContent(c *C) {
 	c.Assert(response.StatusCode, Equals, http.StatusOK)
 
 	buffer1 := bytes.NewReader([]byte("Hello World"))
-	request, err = newTestRequest("PUT", s.testServer.Server.URL+"/partial-content/bar",
-		int64(buffer1.Len()), buffer1, s.testServer.AccessKey, s.testServer.SecretKey)
+	request, err = newTestRequest("PUT", getPutObjectURL(s.endPoint, bucketName, "bar"),
+		int64(buffer1.Len()), buffer1, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
 
 	client = http.Client{}
@@ -1059,8 +1059,8 @@ func (s *TestSuiteCommon) TestPartialContent(c *C) {
 	c.Assert(response.StatusCode, Equals, http.StatusOK)
 
 	// Prepare request
-	request, err = newTestRequest("GET", s.testServer.Server.URL+"/partial-content/bar",
-		0, nil, s.testServer.AccessKey, s.testServer.SecretKey)
+	request, err = newTestRequest("GET", getGetObjectURL(s.endPoint, bucketName, "bar"),
+		0, nil, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
 	request.Header.Add("Range", "bytes=6-7")
 
@@ -1100,7 +1100,7 @@ func (s *TestSuiteCommon) TestListObjectsHandlerErrors(c *C) {
 	response, err = client.Do(request)
 	c.Assert(err, IsNil)
 	// validating the error response.
-	verifyError(c, response, "InvalidArgument", "Argument maxKeys must be an integer between 0 and 2147483647.", http.StatusBadRequest)
+	verifyError(c, response, "InvalidArgument", "Argument maxKeys must be an integer between 0 and 2147483647", http.StatusBadRequest)
 }
 
 // TestPutBucketErrors - request for non valid bucket operation
@@ -1148,7 +1148,7 @@ func (s *TestSuiteCommon) TestPutBucketErrors(c *C) {
 
 	response, err = client.Do(request)
 	c.Assert(err, IsNil)
-	verifyError(c, response, "NotImplemented", "A header you provided implies functionality that is not implemented.", http.StatusNotImplemented)
+	verifyError(c, response, "NotImplemented", "A header you provided implies functionality that is not implemented", http.StatusNotImplemented)
 }
 
 func (s *TestSuiteCommon) TestGetObjectLarge10MiB(c *C) {
@@ -1289,7 +1289,7 @@ func (s *TestSuiteCommon) TestGetPartialObjectMisAligned(c *C) {
 	bucketName := getRandomBucketName()
 	// HTTP request to create the bucket.
 	request, err := newTestRequest("PUT", getMakeBucketURL(s.endPoint, bucketName),
-		0, nil, s.testServer.AccessKey, s.testServer.SecretKey)
+		0, nil, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
 
 	client := http.Client{}
@@ -1321,7 +1321,7 @@ func (s *TestSuiteCommon) TestGetPartialObjectMisAligned(c *C) {
 	objectName := "test-big-file"
 	// HTTP request to upload the object.
 	request, err = newTestRequest("PUT", getPutObjectURL(s.endPoint, bucketName, objectName),
-		int64(buf.Len()), buf, s.testServer.AccessKey, s.testServer.SecretKey)
+		int64(buf.Len()), buf, s.accessKey, s.secretKey)
 	c.Assert(err, IsNil)
 
 	client = http.Client{}
@@ -1351,7 +1351,7 @@ func (s *TestSuiteCommon) TestGetPartialObjectMisAligned(c *C) {
 	for _, t := range testCases {
 		// HTTP request to download the object.
 		request, err = newTestRequest("GET", getGetObjectURL(s.endPoint, bucketName, objectName),
-			0, nil, s.testServer.AccessKey, s.testServer.SecretKey)
+			0, nil, s.accessKey, s.secretKey)
 		c.Assert(err, IsNil)
 		// Get partial content based on the byte range set.
 		request.Header.Add("Range", "bytes="+t.byteRange)
@@ -1584,8 +1584,8 @@ func (s *TestSuiteCommon) TestGetObjectRangeErrors(c *C) {
 	// HTTP request to download the object.
 	request, err = newTestRequest("GET", getGetObjectURL(s.endPoint, bucketName, objectName),
 		0, nil, s.accessKey, s.secretKey)
-	// invalid byte range set.
-	request.Header.Add("Range", "bytes=7-6")
+	// Invalid byte range set.
+	request.Header.Add("Range", "bytes=13-")
 	c.Assert(err, IsNil)
 
 	client = http.Client{}
@@ -1593,7 +1593,7 @@ func (s *TestSuiteCommon) TestGetObjectRangeErrors(c *C) {
 	response, err = client.Do(request)
 	c.Assert(err, IsNil)
 	// expected to fail with "InvalidRange" error message.
-	verifyError(c, response, "InvalidRange", "The requested range cannot be satisfied.", http.StatusRequestedRangeNotSatisfiable)
+	verifyError(c, response, "InvalidRange", "The requested range is not satisfiable", http.StatusRequestedRangeNotSatisfiable)
 }
 
 // TestObjectMultipartAbort - Test validates abortion of a multipart upload after uploading 2 parts.
@@ -1878,7 +1878,7 @@ func (s *TestSuiteCommon) TestObjectMultipartListError(c *C) {
 	c.Assert(err, IsNil)
 	// Since max-keys parameter in the ListMultipart request set to invalid value of -2,
 	// its expected to fail with error message "InvalidArgument".
-	verifyError(c, response4, "InvalidArgument", "Argument maxParts must be an integer between 1 and 10000.", http.StatusBadRequest)
+	verifyError(c, response4, "InvalidArgument", "Argument max-parts must be an integer between 0 and 2147483647", http.StatusBadRequest)
 }
 
 // TestObjectValidMD5 - First uploads an object with a valid Content-Md5 header and verifies the status,

--- a/test-utils_test.go
+++ b/test-utils_test.go
@@ -187,8 +187,8 @@ func newTestRequest(method, urlStr string, contentLength int64, body io.ReadSeek
 			return nil, e
 		}
 		hashedPayload = hex.EncodeToString(sum256(payloadBytes))
-		md5base64 := base64.StdEncoding.EncodeToString(sumMD5(payloadBytes))
-		req.Header.Set("Content-Md5", md5base64)
+		md5Base64 := base64.StdEncoding.EncodeToString(sumMD5(payloadBytes))
+		req.Header.Set("Content-Md5", md5Base64)
 	}
 	req.Header.Set("x-amz-content-sha256", hashedPayload)
 

